Firestore/core/src/firebase/firestore/model/maybe_document.h

```diff
@@ -51,61 +54,114 @@ class MaybeDocument {
     UnknownDocument,
   };
 
```

Comment:
One very interesting property of this class is how copying derived class' objects "just works" without slicing. Perhaps:
* note that in the class comment?
* add checks to ensure slicing cannot occur in the future? Something like `static_assert(sizeof(*this) == sizeof(MaybeDocument)` added to each derived class?